### ISSUE TITLE
MacOS X Option/Alt key in Vim/NeoVim

### ISSUE THREAD

**Issue by daniel-samson** [alacritty/alacritty#4048]
When using neovim with alacritty, the map bindings that have the option key are not working.

### Example:

add the following to your vimrc or init.vim config
```vim
" Project Browsing
nnoremap <M-P> :e .<cr>
inoremap <M-P> <Esc>:e .<cr>
vnoremap <M-P> <Esc>:e .<cr>
```

### System

OS: macOS 10.15.6 (19G73)
Version: alacritty 0.4.3 (c723247)
Shell: [oh my zsh](https://github.com/ohmyzsh/ohmyzsh) and [starship](https://starship.rs/)

### Logs

Keyboard and bindings: 
[2020-07-29 04:44:35.621908000] [INFO] glutin event: WindowEvent { window_id: WindowId(Id(140460401168496)), event: KeyboardInput { device_id: DeviceId(DeviceId), input: KeyboardInput { scancode: 35, state: Pressed, virtual_keycode: Some(P), modifiers: ALT }, is_synthetic: false } }

**Comment by chrisduerr** [comment]
Have you disabled `alt_send_esc` in your config?

**Comment by daniel-samson** [comment]
This is what is in my config
```toml
# Send ESC (\x1b) before characters when alt is pressed.
alt_send_esc: true
```

i have tried to both true and false. It doesn't seem to effect the option key.

**Comment by chrisduerr** [comment]
It certainly does affect the option key. If you have this option enabled, vim won't be able to detect your option key. So everything's working as intended.

Have you tried setting this option to `false` and restarting Alacritty?

**Comment by daniel-samson** [comment]
OK ... I have just tried setting it to false and restarting Alacritty. My Bindings still does not seem to work. I have the same set up on ITerm2. It seems to work on ITerm2. Is there any else that I can do to narrow down the issue?

**Comment by chrisduerr** [comment]
Try pressing the following while in a shell without tmux or anything else running: <kbd>Ctrl</kbd>+<kbd>V</kbd> -> <kbd>Option</kbd>+<kbd>P</kbd>.

If that prints `^[p` with `^[` being inverse, then everything is good from Alacritty's perspective. It should only do that with `alt_send_esc: true`.

Sorry for stating that you should set this to `false` previously, that was incorrect. It should be set to `true`. Because you want the alt key to print the special escape sequences.

Also, could you please try mapping to `<M-p>` instead of `<M-P>`?

**Comment by daniel-samson** [comment]
Can i just clarify what the "->" means?


### alt_send_esc: true
Results:
When I press Ctrl+V once, nothing happens.
When I press Ctrl+V twice, i get **^V**
When I press Option + p, i get **π**.
When i press Option + Shift + P, I get **∏**

### alt_send_esc: false
the results are the same.

Also, I have both <M-p> and <m-p> mapped and neither work in alacritty. [You can see my dot file(s) here](https://github.com/daniel-samson/neovim-macsos/blob/master/.config/nvim/init.vim). My bindings where <A-....> but i changed them to <M...> to see if that would fix the issue. Perhaps i should change them back.

**Comment by chrisduerr** [comment]
>  When I press Option + p, i get π.

There's your problem. Your keyboard layout sends different keys when option is pressed. So if Option + p sends `π`, then that's what you have to bind to.

**Comment by aemonk** [comment]
> It certainly does affect the option key. If you have this option enabled, vim won't be able to detect your option key. So everything's working as intended.
> 
> Have you tried setting this option to `false` and restarting Alacritty?

Just set the `alt_send_esc: false` and every special character with ALT worked fine.

**Comment by Sleepful** [comment]
Yeah, `option` is not recognized as `alt` inside Alacritty. Setting `alt_send_esc: true` or `alt_send_esc: false` doesn't make a difference.

[Kitty terminal has this option, for example.](https://sw.kovidgoyal.net/kitty/conf.html#opt-kitty.macos_option_as_alt)

**Comment by Sleepful** [comment]
A work-around might be [this comment](https://github.com/alacritty/alacritty/issues/62#issuecomment-347552058)

**Comment by laggardkernel** [comment]
> Can i just clarify what the "->" means?
> 
> ### alt_send_esc: true
> Results:
> When I press Ctrl+V once, nothing happens.
> When I press Ctrl+V twice, i get **^V**
> When I press Option + p, i get **π**.
> When i press Option + Shift + P, I get **∏**
> 
> ### alt_send_esc: false
> the results are the same.

The <kbd>Meta</kbd> could be implemented as 8 bit single char, or prepending with `ESC`. This setting `alt_send_esc` switches between this behavior.

**Comment by craigmac** [comment]
> Can i just clarify what the "->" means?
> 
> ### alt_send_esc: true
> Results: When I press Ctrl+V once, nothing happens. When I press Ctrl+V twice, i get **^V** When I press Option + p, i get **π**. When i press Option + Shift + P, I get **∏**
> 
> ### alt_send_esc: false
> the results are the same.
> 
> Also, I have both and mapped and neither work in alacritty. [You can see my dot file(s) here](https://github.com/daniel-samson/neovim-macsos/blob/master/.config/nvim/init.vim). My bindings where <A-....> but i changed them to <M...> to see if that would fix the issue. Perhaps i should change them back.

Tried the same here, using either option for alt sends escape does not work on macOS.

**Comment by sultanahamer** [comment]
I have struggled with this issue. I had to map `<M-q>` which is option(alt) + q in neovim and was not able to. 


Solution is give below. Read complete TLDR before following the links if you are in hurry. Else you can go through Journey below.

**TLDR**
this link https://stackoverflow.com/questions/11876485/how-to-disable-generating-special-characters-when-pressing-the-alta-optiona talks about the "macos's extra non developer friendly magic" and also its solutions. Try those and they should be alright of a fix. I had one more issue, external keyboard with pc's ctrl, windows, alt layout compared to mac's ctrl, option, command explained in story below for which I used this application called Karabiner Elements https://github.com/pqrs-org/KE-complex_modifications to map the keys as per mac layout to make up for my mac's muscle memory

This application has a rule to remove the magic mentioned above as well solving both the issues and the rule is well named as "Map Option to soft-meta for Alacritty" You can download the rule from https://ke-complex-modifications.pqrs.org/ after installing Karabiner elements


**Journey**
After a day of searching all across, what I found is that OSX maps option + alphabets to mappings like below
```
abcdefghijklmnopqrstuvwxyz
å∫ç∂´ƒ©˙ˆ∆˚¬µ˜øπœ®ß†¨√∑≈¥Ω
```

I figured this issue by finding this https://stackoverflow.com/questions/11876485/how-to-disable-generating-special-characters-when-pressing-the-alta-optiona

There were multiple solutions given. But I could not apply the keyboard layout that was shown there as it was not appearing in the keyboard layout even after following the process mentioned. Good that I failed with keylayout approach. I found another solution.

While I was dealing with this issue, I also had another. I was not using a external apple keyboard but a mechanical one where the key layout in bottom line is  ctrl, super key(windows), meta(alt, option) where as for macos keyboard it is ctrl, option(alt), command(super)

This is a bigger user experience issue as I am used to mac os keyboards and muscle memory of command + space or command + tab always is a issue. So, I used macos's modifier keys present in "System preferences" to map option to command and command to option key

iTerm used to identify alt or meta press properly using the option of "left option key as Esc+" from its " Preferences -> profile -> keys -> general" but once I did the maco's modifier keys to map option as command and viceversa, iTerm stopped working identifying alt press as Esc+

With that, I recognized that modifier keys option from mac are also doing some magic. Started searching for alternative for modifier keys and stumbled upon Karabiner elements https://github.com/pqrs-org/KE-complex_modifications

Using this I assigned for my external mechanical keyboard option as command vice versa and iterm started working usual again with its alt as Esc+ option

After this, I also found it has a section for custom rule to remove the magic mentioned above as well . The rule is well named as "Map Option to soft-meta for Alacritty" You can download the rule from https://ke-complex-modifications.pqrs.org/ after installing Karabiner elements

**Comment by qmmp123** [comment]
So. I solved this problem. I had to create keyboard layout without greek letters. To create new layout you can use ukelele app

**Comment by louielyl** [comment]
For iTerm2 users, I solved the issue by going to `Profiles` -> `Keys` -> Change `Left Option key` & `Right Option key` from `normal` to `meta`, hope it helps!

**Comment by kaidev0711** [comment]
window.option_as_alt: true
-- Resize window using <shift> arrow keys
keymap("n", "<A-Up>", "<cmd>resize +2<CR>")
keymap("n", "<A-Down>", "<cmd>resize -2<CR>")
keymap("n", "<A-Left>", "<cmd>vertical resize -2<CR>")
keymap("n", "<A-Right>", "<cmd>vertical resize +2<CR>")
